pkg/statistics/histogram.go

```diff
@@ -1528,10 +1521,83 @@ func MergePartitionHist2GlobalHist(sc *stmtctx.StatementContext, hists []*Histog
 				sum += buckets[i-1].Count
 				bucketNDV += buckets[i-1].NDV
 			}
-			merged, err := mergePartitionBuckets(sc, buckets[i:r])
+			res, err := currentLeftMost.Compare(sc.TypeCtx(), buckets[i].lower, collate.GetBinaryCollator())
+			if err != nil {
+				return nil, err
+			}
+			if res > 0 {
+				currentLeftMost = buckets[i].lower
+			}
+
+			// Iterate possible overlapped ones.
+			// We need to re-sort this part.
+			mergeBuffer = mergeBuffer[:0]
+			cutAndFixBuffer = cutAndFixBuffer[:0]
+			origI := i
+			for ; i > 0; i-- {
+				res, err := buckets[i-1].upper.Compare(sc.TypeCtx(), currentLeftMost, collate.GetBinaryCollator())
+				if err != nil {
+					return nil, err
+				}
+				// If buckets[i-1].upper < currentLeftMost, this bucket has no overlap with current merging one. Break it.
+				if res < 0 {
+					break
+				}
+				// Now the bucket[i-1].upper > currentLeftMost, they are overlapped.
+				res, err = buckets[i-1].lower.Compare(sc.TypeCtx(), currentLeftMost, collate.GetBinaryCollator())
+				if err != nil {
+					return nil, err
+				}
+				// If buckets[i-1].lower >= currentLeftMost, this bucket is totally inside. So it can be totoally merged.
+				if res >= 0 {
+					sum += buckets[i-1].Count
+					bucketNDV += buckets[i-1].NDV
+					mergeBuffer = append(mergeBuffer, buckets[i-1])
+					continue
+				}
+				// Now buckets[i-1].lower < currentLeftMost < buckets[i-1].upper
+				overlapping := calcFraction4Datums(buckets[i-1].lower, buckets[i-1].upper, currentLeftMost)
+				overlappedCount := int64(float64(buckets[i-1].Count) * overlapping)
+				overlappedNDV := int64(float64(buckets[i-1].NDV) * overlapping)
+				sum += overlappedCount
+				bucketNDV += overlappedNDV
+				buckets[i-1].Count -= overlappedCount
+				buckets[i-1].NDV -= overlappedNDV
+
+				// Cut it.
+				cutBkt := newbucket4MergingForRecycle()
+				buckets[i-1].upper.Copy(cutBkt.upper)
+				currentLeftMost.Copy(cutBkt.lower)
+				cutBkt.Count = overlappedCount
+				cutBkt.NDV = overlappedNDV
+				mergeBuffer = append(mergeBuffer, cutBkt)
+				cutAndFixBuffer = append(cutAndFixBuffer, cutBkt)
+
+				currentLeftMost.Copy(buckets[i-1].upper)
+				buckets[i-1].Repeat = 0
+			}
+			var merged *bucket4Merging
+			if len(cutAndFixBuffer) == 0 {
+				merged, err = mergePartitionBuckets(sc, buckets[i:r])
+			} else {
+				// The buckets in buckets[i:origI] needs a re-sort.
+				err = sortBucketsByUpperBound(sc.TypeCtx(), buckets[i:origI])
+				if err != nil {
+					return nil, err
+				}
+				// The content in the merge buffer don't need a re-sort since we just fix some lower bound for them.
+				mergeBuffer = append(mergeBuffer, buckets[origI:r]...)
+				merged, err = mergePartitionBuckets(sc, mergeBuffer)
+				for _, bkt := range cutAndFixBuffer {
+					releasebucket4MergingForRecycle(bkt)
+				}
+				i = origI
```

Comment:
I think this `origI` will be wrong in some cases.
The "totally inside" and "overlap and cut" cases can happen together.
We need this `origI` to restore the `i` to its position before "overlap and cut", but currently `origI` is saved before handling "totally inside".
I think we need to save `origI` before the `continue`.

Reply:
It's set at line#1510, which should be what you mean.

Reply:
Let me add an example:
| bucket idx   |      [lower,upper]      |  comments |
|:----------|:-------------|:------|
| ... |  ... |  |
|x+1|[10,35]| partial overlap (similar to x+3)|
|x+2|[35,40]| totally inside (this buckets should be directly added to `mergeBuffer`)|
|x+3|[20,45]| partial overlap <br/> (it should be cut, this buckets becomes [20,30], and a new bucket [30,45] added to `mergeBuffer`|
|x+4|[30,50]| threshold met when this bucket is added to `sum`|
| ... |  ... |  |

In this case:
- How should the "re-sort" happen?
- What should `origI`, which is also `r` for the next iteration be?